File: php/src/Items/BackstagePassItem.php
<?php

declare(strict_types=1);

namespace GildedRose\Items;

use GildedRose\Item;
use GildedRose\Items\Abstract\AbstractItem;
use GildedRose\Items\Interface\ItemInterface;

class BackstagePassItem extends AbstractItem implements ItemInterface
{
    /**
     * @param Item $item
     *
     * @return void
     */
    public function updateQuality(Item $item): void
    {
        match(true) {
            $item->sellIn <= 0 => $this->decreaseQuality($item, $item->quality),
            $item->sellIn <= 5 => $this->increaseQuality($item, 3),
            $item->sellIn <= 10 => $this->increaseQuality($item, 2),
            default => $this->increaseQuality($item)
        };
    }
}

File: ruby/items/backstage_pass_item.rb
require_relative 'normal_item'

class BackstagePassItem < NormalItem
  def spend_day_in_shop
    decrease_sell_in_day
    return item.quality = 0 if item.sell_in.negative?

    increase_quality
    increase_quality if item.sell_in < 10
    increase_quality if item.sell_in < 5

    keep_quality_upper_limit_in_bounds

    self
  end
end
